File: src/main/java/es/developer/achambi/coreframework/ui/BaseFragment.java
package es.developer.achambi.coreframework.ui;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * Base fragment, can override onBack event and allows options menu on action bar
 */
public abstract class BaseFragment extends Fragment {
    private int timesViewCreated = 0;
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(getLayoutResource(), container, false);
        timesViewCreated++;
        return root;
    }



    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if( !isViewRecreated() ) {
            onViewSetup(view, savedInstanceState);
        }
    }

    public boolean isViewRecreated() {
        return timesViewCreated > 1;
    }

    public abstract int getLayoutResource();
    public abstract void onViewSetup(View view, @Nullable Bundle savedInstanceState);
}

File: src/main/java/es/developer/achambi/coreframework/ui/BaseFragment.kt
package es.developer.achambi.coreframework.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseFragment : androidx.fragment.app.Fragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(layoutResource, container, false)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        savedInstanceState?.let {
            onRestoreInstanceState(it)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        onViewSetup(view)
        if(savedInstanceState == null) {
            onDataSetup()
        }
    }

    /**
     * Called whenever the view is created, meant to setup any view state that is required, data
     * that is not static wont be accessed at this step
     */
    abstract fun onViewSetup(view: View)

    /**
     * Called one per fragment creation, won't be called if the fragment is recreated, this step is
     * meant to request data that is going to be needed to display the fragment initial state
     */
    open fun onDataSetup(){}

    /**
     * Called whenever the fragment state needs to be recreated, won't be called if there's no
     * available bundle
     */
    open fun onRestoreInstanceState(savedInstanceState: Bundle){}

    abstract val layoutResource: Int
}
